File: rust/src/bin/p0048.rs
// Project Euler: Problem 48

euler::run_solver!(48);

fn solve() -> String {
    compute(1000)
}

fn compute(upper: u64) -> String {
    use euler::math;

    debug_assert!(upper > 0);

    let m = 10_u64.pow(10);
    let ans = (1..=upper)
        .filter_map(|x| {
            if x % 10 != 0 {
                Some(math::powmod(x, x, m))
            } else {
                None
            }
        })
        .sum::<u64>();

    format!("{:010}", ans % m)
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0047_10() {
        assert_eq!(compute(10), "0405071317");
    }

    #[test]
    fn p0047_1000() {
        assert_eq!(compute(1000), "9110846700");
    }
}

File: ocaml/src/p0048.ml
(* Project Euler: Problem 48 *)

(*
 * Here is an information about modular exponentiation.
 *  - https://en.wikipedia.org/wiki/Modular_exponentiation
 *)

open Core

let compute upper =
  let modulus = Int.pow 10 10 in
  List.range 1 upper ~stop:`inclusive
  |> List.filter ~f:(fun n -> n mod 10 <> 0)
  |> List.map ~f:(fun n -> Euler.Math.powmod n n modulus)
  |> List.sum (module Int) ~f:Fn.id
  |> Fun.flip ( % ) modulus
  |> Printf.sprintf "%010d"
;;

let solve () = compute 1_000

(* Test *)

let%test_unit "10" = [%test_eq: string] (compute 10) "0405071317"
let%test_unit "1000" = [%test_eq: string] (compute 1_000) "9110846700"
